lib/clib/CCorresLemmas.thy

```diff
@@ -1040,119 +1040,211 @@ lemma ccorres_Guard_True_Seq:
+        apply fastforce
+       using nf apply (fastforce simp: no_fail_def)
+      apply (fastforce dest!: EHOther)
+     apply fastforce
     apply clarsimp
-    apply (prop_tac "G s'")
+    apply (prop_tac "G rv' s'")
      apply (fastforce dest: use_valid intro: body_inv)
     apply (prop_tac "inter_t \<in> G'")
      apply (fastforce dest: hoarep_exec[rotated] intro: body_inv)
+    apply (drule_tac x=rv' in spec)
     apply (drule_tac x=s' in spec)
+    apply (prop_tac "rrel rv' (xf inter_t)")
+     apply (fastforce dest: hoarep_exec[OF _ _ setter_rrel, rotated])
     apply (elim impE)
-     apply (drule_tac s'=inter_t and r1="xf t'" in ccorresE_gets_the[OF cond_ccorres]; assumption?)
+     apply (frule_tac s'=inter_t and r1=rv' in ccorresE_gets_the[OF setter_ccorres]; assumption?)
        apply (fastforce intro: no_ofail)
       apply (fastforce dest: EHOther)
-     subgoal by (fastforce dest: hoarep_exec intro: setter_inv_cond)
-    apply (prop_tac "xf inter_t = xf t'")
-     apply (fastforce dest: hoarep_exec[rotated] intro: setter_xf_inv)
+     apply (intro conjI)
+           apply fastforce
+          using no_ofail apply (fastforce elim!: no_ofailD)
+         apply fastforce
+        apply (fastforce dest: hoarep_exec[OF _ _ setter_rrel, rotated])
+       apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
+      apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
+     apply (fastforce dest: hoarep_exec[OF _ _ setter_inv_cond, rotated])
     apply (fastforce simp: whileLoop_def intro: whileLoop_results.intros(3))
     done
 
+  have setter_hoarep:
+    "\<And>r s s' n xstate.
+       \<Gamma>\<turnstile>\<^sub>h \<langle>(setter;; While {s'. cond_xf s' \<noteq> 0} (B';; setter)) # hs,s'\<rangle> \<Rightarrow> (n, xstate)
+       \<Longrightarrow> \<Gamma>\<turnstile> {s' \<in> G'. (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s')}
+              setter
+              {s'. (s' \<in> G' \<and> (s, s') \<in> srel \<and> G r s \<and> rrel r (xf s'))
+                   \<and> (cond_xf s' \<noteq> 0 \<longrightarrow> (s' \<in> C' \<and> C r s = Some True))}"
+    apply (insert setter_ccorres)
+    apply (drule_tac x=r in meta_spec)
+    apply (frule_tac s=s in ccorres_to_vcg_gets_the)
+     apply (fastforce intro: no_ofail)
+    apply (insert setter_rrel)
+    apply (drule_tac x=s in meta_spec)
+    apply (drule_tac x=r in meta_spec)
+    apply (rule hoarep_conj_lift_pre_fix)
+     apply (rule hoarep_conj_lift_pre_fix)
+      apply (insert setter_inv_cond)[1]
+      apply (drule_tac x=s in meta_spec)
+      apply (drule_tac x=r in meta_spec)
+      apply (rule_tac Q'="{s' \<in> G'. cond_xf s' \<noteq> 0 \<longrightarrow> s' \<in> C'}" in conseqPost)
+        apply fastforce
+       apply fastforce
+      apply fastforce
+     apply (rule hoarep_conj_lift_pre_fix)
+      apply (simp add: Collect_mono conseq_under_new_pre)
+     apply (simp add: Collect_mono conseq_under_new_pre)
```

Comment:
Does this one work?
```suggestion
     apply (simp add: hoarep_conj_lift_pre_fix Collect_mono conseq_under_new_pre)
```

Reply:
Sadly not. I can get `apply (fastforce intro: hoarep_conj_lift_pre_fix simp add: Collect_mono conseq_under_new_pre)` to work, but it takes a good few seconds.

Reply:
Hm. Even with `intro!: ..`? If not, we could fall back to `rule ..; simp add: ...`

Reply:
`intro!` does help, actually, and the original `simp`s weren't so fast, so this is probably no slower. I'll go with that. Thanks